extensions/PhabBugz/lib/User.pm

```diff
@@ -0,0 +1,203 @@
+
+    delete $params->{fields};
+
+    if ( my $external_accounts =
+        $params->{attachments}{'external-accounts'}{'external-accounts'} )
+    {
+        foreach my $account (@$external_accounts) {
+            next if $account->{type} ne 'bmo';
+            $params->{bug_user_id} = $account->{id};
+            last;
+        }
+    }
+
+    delete $params->{attachments};
+
+    return $params;
+}
+
+# {
+#   "data": [
+#     {
+#       "id": 2,
+#       "type": "USER",
+#       "phid": "PHID-USER-h4aqihzqsnytz7nsegsr",
+#       "fields": {
+#         "username": "phab-bot",
+#         "realName": "Phabricator Automation",
+#         "roles": [
+#           "admin",
+#           "verified",
+#           "approved",
+#           "activated"
+#         ],
+#         "dateCreated": 1512573120,
+#         "dateModified": 1512574523,
+#         "policy": {
+#           "view": "public",
+#           "edit": "no-one"
+#         }
+#       },
+#       "attachments": {
+#         "external-accounts": {
+#           "external-accounts": [
+#             {
+#               "id": "9",
+#               "type": "bmo"
+#             }
+#           ]
+#         }
+#       }
+#     }
+#   ],
+#   "maps": {},
+#   "query": {
+#     "queryKey": null
+#   },
+#   "cursor": {
+#     "limit": 100,
+#     "after": null,
+#     "before": null,
+#     "order": null
+#   }
+# }
+
+sub new_from_query {
+    my ( $class, $params ) = @_;
+    my ($user) = $class->match($params);
+    return $user;
+}
+
+sub match {
+    my ( $class, $params ) = @_;
+
+    # BMO id search takes precedence if bugzilla_ids is used.
+    my $bugzilla_ids = delete $params->{bugzilla_ids};
+    if ($bugzilla_ids) {
+        my $bugzilla_data =
+          $class->get_phab_bugzilla_ids( { ids => $bugzilla_ids } );
+        $params->{phids} = [ map { $_->{phid} } @$bugzilla_data ];
+    }
+
+    return [] if !@{ $params->{phids} };
+
+    # Look for BMO external user id in external-accounts attachment
+    my $data = {
+        constraints => { phids             => $params->{phids} },
+        attachments => { external_accounts => 1 }
+    };
+
+    my $phab_users = [];
+    my $result = request( 'user.search', $data );
+
+    if ( exists $result->{result}{data} && @{ $result->{result}{data} } ) {
+        foreach my $user ( @{ $result->{result}{data} } ) {
+            push @$phab_users, $class->new($user);
+        }
+    }
+
+    return $phab_users;
+}
+
+#################
+#   Accessors   #
+#################
+
+sub bugzilla_user {
+    my ($self) = @_;
+    return undef if !$self->bugzilla_id;
+    return $self->{bugzilla_user} ||=
+      Bugzilla::User->new( { id => $self->bugzilla_id, cache => 1 } );
+}
+
+sub get_phab_bugzilla_ids {
+    my ( $class, $params ) = @_;
+
+    my $memcache = Bugzilla->memcached;
+
+    # Try to find the values in memcache first
+    my @results;
+    my @bugzilla_ids = @{ $params->{ids} };
```

Comment:
How big do we expect `@bugzilla_ids` to get? splice() is very slow for large arrays.